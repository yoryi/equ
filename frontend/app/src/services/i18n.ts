import filesize from "filesize"
import i18n from "i18next"
import _ from "lodash"
import { initReactI18next } from "react-i18next"

import { en } from "../translations"

void i18n.use(initReactI18next).init({
  ns: [`common`, `screens`],
  defaultNS: `screens`,
  resources: {
    en,
  },
  lng: `en`,
  fallbackLng: `en`,
  interpolation: {
    format: (value, format, lng) => {
      if (!format) {
        return value
      }

      return format.split(`,`).reduce((value: string, format: string) => {
        switch (format.trim()) {
          case `localeString`:
            return value.toLocaleString()
          case `enIndefiniteArticle`:
          case `enIndefiniteArticleCapitalized`: {
            if (!lng || lng !== `en`) {
              return value
            }

            const letters = [`a`, `e`, `i`, `o`, `u`, `h`]
            const firstLetter = value.substring(0, 1)
            let article = ``
            if (
              letters.find((letter) => firstLetter.toLowerCase() === letter)
            ) {
              article =
                format === `enIndefiniteArticleCapitalized` ? `An` : `an`
            } else {
              article = format === `enIndefiniteArticleCapitalized` ? `A` : `a`
            }

            return `${article} ${value}`
          }
          case `title`: {
            return value
              .toLowerCase()
              .split(` `)
              .map((word: string) => _.upperFirst(word))
              .join(` `)
          }
          case `lowerCase`: {
            return value.toLowerCase()
          }
          case `fileSize`: {
            return filesize(parseInt(value), { round: 0 })
          }
          default:
            return value
        }
      }, value)
    },
  },
})

export default i18n
