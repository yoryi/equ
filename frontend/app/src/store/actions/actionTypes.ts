export * from "./admin/actionTypes"
export * from "./attachments/actionTypes"
export * from "./auth/actionTypes"
export * from "./follow/actionTypes"
export * from "./journeys/actionTypes"
export * from "./media/actionTypes"
export * from "./notifications/actionTypes"
export * from "./profile/actionTypes"
export * from "./searchAndFilters/actionTypes"
export * from "./settings/actionTypes"
export * from "./shared/actionTypes"
export * from "./student/actionTypes"
export * from "./surveys/actionTypes"
export * from "./university/actionTypes"
