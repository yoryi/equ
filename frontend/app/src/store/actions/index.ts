export * from "./admin"
export * from "./attachments"
export * from "./auth"
export * from "./follow"
export * from "./help"
export * from "./journeys"
export * from "./media"
export * from "./notifications"
export * from "./profile"
export * from "./searchAndFilters"
export * from "./settings"
export * from "./shared"
export * from "./student"
export * from "./surveys"
export * from "./university"
