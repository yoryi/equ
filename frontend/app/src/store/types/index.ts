export * from "./admin"
export * from "./auth"
export * from "./errors"
export * from "./ethnicity"
export * from "./follow"
export * from "./gender"
export * from "./hardship"
export * from "./help"
export * from "./highSchool"
export * from "./interest"
export * from "./notifications"
export * from "./profile"
export * from "./race"
export * from "./search"
export * from "./settings"
export * from "./shared"
export * from "./store"
export * from "./student"
export * from "./survey"
export * from "./university"
export * from "./utils"
