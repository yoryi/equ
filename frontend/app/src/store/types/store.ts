import { AdminReducer } from "../reducers/admin/reducer"
import { AuthReducer } from "../reducers/auth/reducer"
import { FollowReducer } from "../reducers/follow"
import { MediaReducer } from "../reducers/media/reducer"
import { NotificationsReducer } from "../reducers/notifications"
import { ProfileReducer } from "../reducers/profile/reducer"
import { SearchAndFiltersReducer } from "../reducers/searchAndFilters/reducer"
import { SettingsReducer } from "../reducers/settings/reducer"
import { SharedReducer } from "../reducers/shared/reducer"
import { StudentReducer } from "../reducers/student"
import { SurveysReducer } from "../reducers/surveys/reducer"
import { UniversityReducer } from "../reducers/university"

export interface ReduxState {
  auth: AuthReducer
  media: MediaReducer
  notifications: NotificationsReducer
  profile: ProfileReducer
  shared: SharedReducer
  settings: SettingsReducer
  student: StudentReducer
  surveys: SurveysReducer
  university: UniversityReducer
  follow: FollowReducer
  searchAndFilters: SearchAndFiltersReducer
  admin: AdminReducer
}
