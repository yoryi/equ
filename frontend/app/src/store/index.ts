import * as actions from "./actions"
import { initReduxStore } from "./config"
import { store } from "./createStore"

export { actions,initReduxStore, store }
