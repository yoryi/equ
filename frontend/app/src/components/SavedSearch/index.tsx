import classNames from "classnames"
import * as React from "react"
import { useState } from "react"
import toast from "react-hot-toast"
import { useTranslation } from "react-i18next"
import { useDispatch } from "react-redux"
import { Link } from "react-router-dom"

import { ReactComponent as Close } from "../../assets/close-icon-white.svg"
import * as actions from "../../store/actions"
import { Search } from "../../store/types"
import RoundButton, { RoundButtonState } from "../RoundButton"

interface SavedSearchProps {
  className?: string
  search: Search.SavedSearch
  removable?: boolean
}

const SavedSearch: React.VFC<SavedSearchProps> = ({
  className,
  search,
  removable = false,
}) => {
  const dispatch = useDispatch()

  const [buttonState, setButtonState] = useState<RoundButtonState>(
    RoundButtonState.Default,
  )

  const handleDelete = async () => {
    try {
      setButtonState(RoundButtonState.Loading)
      await dispatch(actions.deleteSavedSearch(search))
    } catch (err) {
      if (err.message) {
        toast.error(err.message)
      }

      setButtonState(RoundButtonState.Error)
    }
  }

  const { t } = useTranslation()

  return (
    <div
      className={classNames(
        `d-flex justify-content-between align-items-center`,
        className,
      )}
    >
      <Link
        to={`/search`}
        className={`d-flex flex-column flex-sm-row align-items-sm-end font-weight-semi-bold`}
        onClick={() => dispatch(actions.setSavedSearch(search.query))}
        role={`search`}
      >
        {search.name}

        <span className={`text-3 ml-0 ml-sm-2`}>
          {`(${t(`common:matches`, { count: search.matching ?? 0 })})`}
        </span>
      </Link>

      {removable && (
        <RoundButton
          variant={`red`}
          state={buttonState}
          onClick={handleDelete}
          role={`delete`}
          secondary
        >
          <Close />
        </RoundButton>
      )}
    </div>
  )
}

export default SavedSearch
