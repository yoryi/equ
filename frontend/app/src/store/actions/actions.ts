/* eslint-disable @typescript-eslint/ban-types */
import { Action } from "redux"

import * as Types from "../types"

export interface PayloadAction<T> extends Action {
  payload: T
}

export interface PromiseAction<T> extends PayloadAction<T> {
  meta: {
    promise: {
      resolve: Function
      reject: Function
    }
  }
  then: Function
  catch: Function
  finally: Function
}

export interface ErrorAction extends Action {
  err: Error
}

export interface APIErrorAction extends Action {
  err: Types.APIError
}
