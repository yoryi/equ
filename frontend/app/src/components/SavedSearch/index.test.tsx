//@ts-ignore
import { promiseMiddleware } from "@adobe/redux-saga-promise"
//Testing
import { act,fireEvent, render } from "@testing-library/react"
import faker from "faker"
import * as React from "react"
//Translations
import { I18nextProvider } from "react-i18next"
//Redux store
import { Provider } from "react-redux"
//Navigation
import { MemoryRouter } from "react-router"
import configureStore from "redux-mock-store"

import i18n from "../../services/i18n"
import * as actions from "../../store/actions"
//Types
import { Search } from "../../store/types"
//Components
import SavedSearch from "./"

describe(`SavedSearch component`, () => {
  faker.seed(1000)

  const mockStore = configureStore([promiseMiddleware])
  const store = mockStore()

  let search: Search.SavedSearch
  beforeEach(() => {
    search = {
      id: faker.random.number(10000),
      name: faker.lorem.words(2),
      matching: faker.random.number({ min: 100, max: 1000 }),
      query: faker.lorem.words(2),
    }
  })

  afterEach(() => {
    store.clearActions()
  })

  const wrapper: React.FC = ({ children }) => (
    <MemoryRouter>
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>{children}</Provider>
      </I18nextProvider>
    </MemoryRouter>
  )

  describe(`should render and match the snapshot`, () => {
    it(`non removable`, () => {
      const { container } = render(<SavedSearch search={search} />, {
        wrapper,
      })

      expect(container.firstChild).toMatchSnapshot()
    })

    it(`removable`, () => {
      const { container } = render(<SavedSearch search={search} removable />, {
        wrapper,
      })

      expect(container.firstChild).toMatchSnapshot()
    })

    it(`no matching count`, () => {
      search = {
        ...search,
        matching: null,
      }

      const { container } = render(<SavedSearch search={search} removable />, {
        wrapper,
      })

      expect(container.firstChild).toMatchSnapshot()
    })
  })

  it(`should execute query`, () => {
    const { getByRole } = render(<SavedSearch search={search} />, {
      wrapper,
    })

    act(() => {
      fireEvent.click(getByRole(`search`))
    })

    expect(store.getActions()).toContainEqual(
      actions.setSavedSearch(search.query),
    )
  })

  describe(`should attempt to delete saved search`, () => {
    it(`success`, () => {
      const action = actions.deleteSavedSearch(search)

      const { getByRole } = render(<SavedSearch search={search} removable />, {
        wrapper,
      })

      act(() => {
        fireEvent.click(getByRole(`delete`))

        store
          .getActions()
          .find(({ type }) => type === action.type)
          ?.meta.promise.resolve()
      })

      expect(
        store.getActions().find(({ type }) => type === action.type).payload,
      ).toEqual(action.payload)
    })
  })
})
