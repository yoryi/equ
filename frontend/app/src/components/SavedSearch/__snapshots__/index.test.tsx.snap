// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedSearch component should render and match the snapshot no matching count 1`] = `
<div
  class="d-flex justify-content-between align-items-center"
>
  <a
    class="d-flex flex-column flex-sm-row align-items-sm-end font-weight-semi-bold"
    href="/search"
    role="search"
  >
    nisi qui
    <span
      class="text-3 ml-0 ml-sm-2"
    >
      (0 matches)
    </span>
  </a>
  <button
    class="button secondary red"
    role="delete"
  >
    <div
      class="content loader exited"
    >
      <div
        aria-busy="true"
        class=""
      >
        <svg
          aria-label="audio-loading"
          fill="#c91c0d"
          height="24"
          viewBox="0 0 120 30"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="15"
            cy="15"
            r="15"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="15"
              repeatCount="indefinite"
              to="15"
              values="15;9;15"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="1"
              repeatCount="indefinite"
              to="1"
              values="1;.5;1"
            />
          </circle>
          <circle
            attributeName="fillOpacity"
            cx="60"
            cy="15"
            from="1"
            r="9"
            to="0.3"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="9"
              repeatCount="indefinite"
              to="9"
              values="9;15;9"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="0.5"
              repeatCount="indefinite"
              to="0.5"
              values=".5;1;.5"
            />
          </circle>
          <circle
            cx="105"
            cy="15"
            r="15"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="15"
              repeatCount="indefinite"
              to="15"
              values="15;9;15"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="1"
              repeatCount="indefinite"
              to="1"
              values="1;.5;1"
            />
          </circle>
        </svg>
      </div>
    </div>
    <div
      class="content entered"
    >
      <svg>
        close-icon-white.svg
      </svg>
    </div>
    <div
      class="content success exited"
    >
      <svg>
        check.svg
      </svg>
    </div>
    <div
      class="content exited"
    >
      <svg>
        close-icon-white.svg
      </svg>
    </div>
  </button>
</div>
`;

exports[`SavedSearch component should render and match the snapshot non removable 1`] = `
<div
  class="d-flex justify-content-between align-items-center"
>
  <a
    class="d-flex flex-column flex-sm-row align-items-sm-end font-weight-semi-bold"
    href="/search"
    role="search"
  >
    quia fugit
    <span
      class="text-3 ml-0 ml-sm-2"
    >
      (649 matches)
    </span>
  </a>
</div>
`;

exports[`SavedSearch component should render and match the snapshot removable 1`] = `
<div
  class="d-flex justify-content-between align-items-center"
>
  <a
    class="d-flex flex-column flex-sm-row align-items-sm-end font-weight-semi-bold"
    href="/search"
    role="search"
  >
    ea officiis
    <span
      class="text-3 ml-0 ml-sm-2"
    >
      (865 matches)
    </span>
  </a>
  <button
    class="button secondary red"
    role="delete"
  >
    <div
      class="content loader exited"
    >
      <div
        aria-busy="true"
        class=""
      >
        <svg
          aria-label="audio-loading"
          fill="#c91c0d"
          height="24"
          viewBox="0 0 120 30"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="15"
            cy="15"
            r="15"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="15"
              repeatCount="indefinite"
              to="15"
              values="15;9;15"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="1"
              repeatCount="indefinite"
              to="1"
              values="1;.5;1"
            />
          </circle>
          <circle
            attributeName="fillOpacity"
            cx="60"
            cy="15"
            from="1"
            r="9"
            to="0.3"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="9"
              repeatCount="indefinite"
              to="9"
              values="9;15;9"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="0.5"
              repeatCount="indefinite"
              to="0.5"
              values=".5;1;.5"
            />
          </circle>
          <circle
            cx="105"
            cy="15"
            r="15"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="15"
              repeatCount="indefinite"
              to="15"
              values="15;9;15"
            />
            <animate
              attributeName="fillOpacity"
              begin="0s"
              calcMode="linear"
              dur="0.8s"
              from="1"
              repeatCount="indefinite"
              to="1"
              values="1;.5;1"
            />
          </circle>
        </svg>
      </div>
    </div>
    <div
      class="content entered"
    >
      <svg>
        close-icon-white.svg
      </svg>
    </div>
    <div
      class="content success exited"
    >
      <svg>
        check.svg
      </svg>
    </div>
    <div
      class="content exited"
    >
      <svg>
        close-icon-white.svg
      </svg>
    </div>
  </button>
</div>
`;
